### ISSUE TITLE
Query about semseg domain in pre-training

### ISSUE THREAD

**Issue by Chianghui-Wong** [EPFL-VILAB/MultiMAE#8]
Hi, I have successful made the pesudo labels and trained ‘rgb’ in/out-domain multimae model.

But when I trained model with 'rgb-semseg' in/out-domain, I met an error in multimae/input_adapters.py line 232

```python
# Create patches [B, C, H, W] -> [B, (H*W), C]
x_patch = rearrange(self.proj(x), 'b d nh nw -> b (nh nw) d')
```
The full log is [log.txt](https://github.com/EPFL-VILAB/MultiMAE/files/8670966/log.txt).
x.size() is [batchsize, 64, 56, 56] before line 232.
I can't find out what's wrong.

What's more, I don't know why the pseudo semeg label image  resize into 1/4 (that is 224\*224->56\*56)in utils/datasets.py line 105
```python
# Convert to Tensor
for task in task_dict:
    if task in ['depth']:
        img = torch.Tensor(np.array(task_dict[task]) / 2 ** 16)
        img = img.unsqueeze(0)  # 1 x H x W
    elif task in ['rgb']:
        img = TF.to_tensor(task_dict[task])
        img = TF.normalize(img, mean=self.rgb_mean, std=self.rgb_std)
    elif task in ['semseg', 'semseg_coco']:
        # TODO: add this to a config instead
        # Rescale to 0.25x size (stride 4)
        scale_factor = 0.25
        img = task_dict[task].resize((int(self.input_size * scale_factor), int(self.input_size * scale_factor)))
        # Using pil_to_tensor keeps it in uint8, to_tensor converts it to float (rescaled to [0, 1])
        img = TF.pil_to_tensor(img).to(torch.long).squeeze(0)
```
and then use nn.Conv2d in multimae/input_adapters.py line  198
```python
if self.interpolate_class_emb:
    self.proj = nn.Sequential(
        nn.Upsample(scale_factor=(1 / self.P_H, 1 / self.P_W),
                    mode='bilinear'),  # Actually a downsample operation
        nn.Conv2d(in_channels=self.dim_class_emb, out_channels=self.dim_tokens,
                    kernel_size=1, stride=1),
    )
else:
    self.proj = nn.Conv2d(
        in_channels=self.dim_class_emb, out_channels=self.dim_tokens,
        kernel_size=(self.P_H, self.P_W), stride=(self.P_H, self.P_W)
    )
)
```

Thank you for any help.

**Comment by Chianghui-Wong** [comment]
And 'rgb-depth' domain is OK.

I only use the 50 classes of imagenet, which is imagenet-50.

**Comment by dmizr** [comment]
Hi @Chianghui-Wong ,

In our experiments, we downsample the semantic segmentation input to 1/4th the size of RGB and depth and use a projection layer with patches of size 4x4 with stride 4. This keeps the number of segmentation patches constant while making this layer much more computationally efficient. If we kept it at full resolution, this convolutional layer would have approx. 12M params for 64-dimensional class embeddings, as opposed to less than 1M params with this adjustment. 

From the log that you sent, it seems like you the segmentation images are indeed resized to 1/4th the size (56x56 instead of 224x224), but the projection layer (line 13 of your log) still has a kernel size of 16x16 and stride of 16, hence the size mismatch and the error. If you change this layer to a kernel size of 4 and stride of 4, then it should hopefully work.

Hope this helps!

**Comment by roman-bachmann** [comment]
Hi @Chianghui-Wong, 

Another potential source of this error could be an out-of-bounds class index for the nn.Embedding layer in [SemSegInputAdapter](https://github.com/EPFL-VILAB/MultiMAE/blob/main/multimae/input_adapters.py#L195). By default, it assumes 133 classes for COCO (so an invalid class index like 255 would not be handled).

Best, Roman
